### DISCUSSION TITLE
Transitions to ancestor's sibling state allowed in scxml but not alowed in xstate. Why?

### DISCUSSION THREAD

**Discussion by dimsuz** [statelyai/xstate#1647] [Q&A] [answered]
Hi!

Scion scxml test framework has this test `hier2.scxml`:

```xml
<scxml>
    <state id="a">
        <state id="a1">
            <transition target="b" event="t"/>
        </state>
        <state id="a2">
        </state>
        <transition target="a2" event="t"/>
    </state>
    <state id="b"/>
</scxml>
```
Notice that `a.a1` has the transition targeting state `b`. But if I try to create a corresponding machine in `xstate`, it will give me an error that `b` is not a sibling of `a1`. 

I have these questions:

1. I wonder was this a conscious decision or is this some kind of bug (I doubt this). I've searched issues a bit and it seems that instead of allowing this kind of target, `xstate` prefers to encode such transitions using `#machine.state` pattern. Am I right?
2. I also noticed that in `v5`/`next` branch there are these scion tests and they somehow contain this `hier2.scxml/json` test so I guess it must pass in the `next` branch, otherwise it wouldn't be included. If so, does this mean that this restriction has been lifted in `next`? If so, where can I find the resolution algorithm?

I'm trying to port xstate to kotlin and decided to use those scion test-framework tests too, got a failing one, hence the questions :)

**Comment by davidkpiano** [comment, marked as answer]
Transition to an ancestor's sibling state (or any state, rather) is definitely allowed in XState.

The above machine would have:
```js
on: {
  t: '#b'
}
```

And a state with:
```js
states: {
  b: {
    id: 'b'
  }
}
```

This is actually no different from what's happening in SCXML. The only apparent difference with SCXML is that _all transitions are by ID_, by default. Notice how all states have `id` on them.

So to do the same behavior, you can target any state by its ID in exactly the same way (with a `#` sign in front). XState by default has transition targets _relative by default_, and this is for two reasons:

- Pit of success: by making targets relative by default, it promotes having a sensible state hierarchy, where "escaping" the hierarchy would feel unnatural
- Prevents naming conflicts. Instead of manually specifying every ID (as in SCXML), IDs are autogenerated based on their path, so there are no conflicts.

**Reply by dimsuz** [reply]
Awesome! Thank you for clarifying that (and so quickly). I definitely had some vague notion of keys vs ids vs relative-ids, now this picture is clearer!

I will then adjust my "from-scxml" converter accordingly so that it will generate a machine schema which will target by id in these situations (or maybe in all situations as it is an auto-conversion anyways, not something written by a user).

**Reply by davidkpiano** [reply]
By the way, if you want to stick to SCXML's algorithm, I recommend looking at the [XState Python implementation](https://github.com/davidkpiano/xstate-python/blob/master/xstate/algorithm.py) as a guide. XState v4 does things a bit custom (but at least it still passes most of the SCXML tests), and XState v5 is closer to the actual algorithm.

**Reply by dimsuz** [reply]
Yep, that's what I did (you suggested this to me on twitter recently), already ported it for like 90%, the only thing left is actions/context support. Just got `cond` scion tests working today. I've used mozilla's `rhino` library which allows executing JS from Java (→Kotlin), use it to evaluate cond expressions in those scxml tests.

And then I plan to enable more and more scion tests and then maybe refactor to prettify as it's quite crude in some places, non-idiomatic port from python :)
